ml_genn/model.py

```diff
@@ -305,23 +305,23 @@ def evaluate(self, data, labels, time, save_samples=[]):
         if any(i < 0 or i >= n_samples for i in save_samples):
             raise ValueError('one or more invalid save_samples value')
 
+        n_complete = [0] * len(self.outputs)
         n_correct = [0] * len(self.outputs)
         accuracy = [0] * len(self.outputs)
         all_spikes = [[[] for i,_ in enumerate(self.layers)] for s in save_samples]
 
         # Pad number of samples so pipeline can be flushed
-        pipeline_depth = self.calc_pipeline_depth()
-        padded_n_samples = n_samples + (pipeline_depth * self.g_model.batch_size)
+        pipeline_depth = [l.pipeline_depth if hasattr(l, 'pipeline_depth') else 0 for l in self.outputs]
+        padded_n_samples = [n_samples + (depth * self.g_model.batch_size) for depth in pipeline_depth]
 
         # Process batches
         progress = tqdm(total=n_samples)
-        for batch_start in range(0, padded_n_samples, self.g_model.batch_size):
+        for batch_start in range(0, max(padded_n_samples), self.g_model.batch_size):
 
             # If any elements of this batch have data (rather than being entirely pipeline padding)
             if batch_start < n_samples:
```

Comment:
The body of the loops here and in ``evaluate_batched`` are very similar so I wonder if they could be moved into a shared helper method - I _think_ you could pass in the next batch of data for each input (if there is one) and label for each output (if there is one) and the rest could be pretty generic (although I may be missing something)

Reply:
I've refactored `evaluate` to batch data and send off to `evaluate_batched`. Should be a bit cleaner now.